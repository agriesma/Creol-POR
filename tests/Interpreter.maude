***
*** Interpreter.maude -- Test cases for the interpreter.
***
*** Copyright (c) 2007
***
*** This program is free software; you can redistribute it and/or
*** modify it under the terms of the GNU General Public License as
*** published by the Free Software Foundation; either version 2 of the
*** License, or (at your option) any later version.
***
*** This program is distributed in the hope that it will be useful, but
*** WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
*** General Public License for more details.
***
*** You should have received a copy of the GNU General Public License
*** along with this program; if not, write to the Free Software
*** Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA
*** 02111-1307, USA.
***

set show stats off .
set show timing off .
set show advise off .
load interpreter .

mod TEST-TAILCALL is
  pr INTERPRETER .
  op init : -> Configuration .
  eq init =

    < 'Tailcall : Cl | Inh: noInh, Par: 'x, Att: 'x |-> null, Mtds:
	< 'init : Mtdname | Param: noAid, Latt: noSubst, Code: return(emp) > *
	< 'run : Mtdname | Param: noAid, Latt: noSubst, Code:
          if 'less[[ int(0) # 'x ]] th 'x ::= ('minus[[ 'x # int(1) ]]) ;
	    tailcall 'run(emp) el return(emp) fi >,
      Ocnt: 0 >

  main('Tailcall, int(5)) .

endm

red in TEST-TAILCALL : init .

search [1] in TEST-TAILCALL : init =>+
  < ob('Tailcall0) : 'Tailcall | Att: A:Subst, Pr:
    L:Subst, accept(tag(label(ob('Tailcall0), 2))) ; SL:StmList, PrQ: W:MProc, Lcnt: N:Nat >
  conf:Configuration .

rew in TEST-TAILCALL : init .

quit
