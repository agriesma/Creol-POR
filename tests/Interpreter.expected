==========================================
reduce in TEST-TAILCALL : init .
result Configuration: 
< ob('main) : Qu | Dealloc: noDealloc,Ev: noMsg >
 
< 'Tailcall : Cl | Inh: noInh,Par: 'x,Att: 'x |-> null,Mtds:
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: if 'less[[int(0) # 'x]] th
    'x ::= ('minus[['x # int(1)]]) ; tailcall 'run(emp) el return(emp) fi >,
    Ocnt: 0 >
 
< ob('main) : 'Class | Att: noSubst,Pr: noSubst,
'var ::= new 'Tailcall(int(5)),PrQ: noProc,Lcnt: 0 >

==========================================
search [1] in TEST-TAILCALL : init =>+ conf:Configuration 
< ob('Tailcall0) : 'Tailcall | Att: A:Subst,Pr: L:Subst,
accept(tag(label(2))) ; SL:StmList,PrQ: W:MProc,Lcnt: N:Nat >
 .

Solution 1 (state 42)
conf:Configuration --> boundMtd(ob('Tailcall0),('caller |-> ob('Tailcall0),
    'label |-> tag(label(2))),
noAid ::= emp ; if 'less[[int(0) # 'x]] th 'x ::= ('minus[['x # int(1)]]) ;
    tailcall 'run(emp) el return(emp) fi)
 
< ob('Tailcall0) : Qu | Dealloc: noDealloc,Ev: noMsg >
 
< ob('main) : Qu | Dealloc: noDealloc,Ev: noMsg >
 
< 'Tailcall : Cl | Inh: noInh,Par: 'x,Att: 'x |-> null,Mtds:
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: if 'less[[int(0) # 'x]] th
    'x ::= ('minus[['x # int(1)]]) ; tailcall 'run(emp) el return(emp) fi >,
    Ocnt: 1 >
 
< ob('main) : 'Class | Att: noSubst,Pr: noSubst,
'var assign ob('Tailcall0),PrQ: noProc,Lcnt: 0 >

A:Subst --> 'this |-> ob('Tailcall0), 'x |-> int(4)
L:Subst --> noSubst
SL:StmList --> noStm
W:MProc --> 'L0 |-> label(2),
label(2) ?(noAid)
N:Nat --> 3
==========================================
rewrite in TEST-TAILCALL : init .
result Configuration: 
< ob('Tailcall0) : Qu | Dealloc: noDealloc,Ev: noMsg >
 
< ob('main) : Qu | Dealloc: noDealloc,Ev: noMsg >
 
< 'Tailcall : Cl | Inh: noInh,Par: 'x,Att: 'x |-> null,Mtds:
< 'init : Mtdname | Param: noAid,Latt: noSubst,Code: return(emp) > *
< 'run : Mtdname | Param: noAid,Latt: noSubst,Code: if 'less[[int(0) # 'x]] th
    'x ::= ('minus[['x # int(1)]]) ; tailcall 'run(emp) el return(emp) fi >,
    Ocnt: 1 >
 
< ob('Tailcall0) : 'Tailcall | Att: 'this |-> ob('Tailcall0), 'x |-> int(0),Pr:
    idle,PrQ: noProc,Lcnt: 3 >
 
< ob('main) : 'Class | Att: noSubst,Pr: idle,PrQ: noProc,Lcnt: 0 >

Bye.
